File: src/hnsw_build.cpp
#include <hnswlib/hnswlib.h>
#include <numeric>
#include <oneapi/tbb/global_control.h>
#include <oneapi/tbb/info.h>
#include <oneapi/tbb/parallel_for.h>
#include <queue>
#include <vector>
#include <spdlog/spdlog.h>

#include "cmdline.hpp"
#include "dataloader.hpp"
#include "span.hpp"
#include "timer.hpp"

using namespace melee;
template <typename space_t, typename dist_t> void build(HNSWBuildConfig config) {
    typedef std::priority_queue<std::pair<dist_t, hnswlib::labeltype>> ResultType;

    Matrix2D feat = loadMatrix(config.feat_path, config.max_elements);
    space_t space(feat.shape[1]);
    hnswlib::HierarchicalNSW<dist_t> *alg_hnsw{nullptr};
    Timer timer;
    timer.start();
    ALWAYS_ASSERT(!config.feat_path.empty());
    spdlog::info("Loading Data From: {}", config.feat_path);
    spdlog::info("Adding {}M points", config.max_elements / 1e6);

    alg_hnsw = new hnswlib::HierarchicalNSW<dist_t>(
            &space, config.max_elements, config.M, config.ef_construction);

    oneapi::tbb::parallel_for(
            oneapi::tbb::blocked_range<size_t>(0, config.max_elements, feat.shape[0] / 1000),
            [&](const oneapi::tbb::blocked_range<size_t> &r) {
                for (size_t i = r.begin(); i < r.end(); i++) {
                    alg_hnsw->addPoint(feat.get_vec(i), i);
                }
            }, oneapi::tbb::simple_partitioner());

    timer.end();
    spdlog::info("BuildTime={} secs", timer.seconds());
    spdlog::info("Saving Index To: {}", config.index_path);
    alg_hnsw->saveIndex(config.index_path);
}

int main(int argc, char *argv[]) {
    HNSWBuildConfig config;
    config.Init(argc, argv);

    if (config.space == "ip") {
        build<hnswlib::InnerProductSpace, float>(config);
    } else if (config.space == "l2") {
        if (ends_with(config.feat_path, "fbin")) {
            build<hnswlib::L2Space, float>(config);
        } else if (ends_with(config.feat_path, "u8bin")){
            build<hnswlib::L2SpaceI, int>(config);
        }
    } else {
        spdlog::error("Invalid space {}", config.space);
    }
}

File: python/hnsw_build.py
import hnswlib
import argparse
import time
from util import *

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='HNSW build script')
    parser.add_argument('--space', type=str, help='distance space')
    parser.add_argument('--ef_construction', type=int, help='ef construction')
    parser.add_argument('--M', type=int, help='out degree upper bound')
    parser.add_argument('--max_elements', type=int, help='num elements to build index')
    parser.add_argument('--feat_path', type=str, help='feature path')
    parser.add_argument('--index_path', type=str, help='index save path')
    args = parser.parse_args()
    
    print(f"{args=}")
    
    t = time.time()
    feat = load_data(args.feat_path, args.max_elements)
    print(f"load feat in {time.time() - t} secs")
    
    num_elements = feat.shape[0]
    dim = feat.shape[1]
    ids = np.arange(num_elements)
    
    t = time.time()
    p = hnswlib.Index(space = args.space, dim = dim)
    p.init_index(max_elements = num_elements, ef_construction = args.ef_construction, M = args.M)
    p.add_items(feat, ids)
    
    print(f"build index in {time.time() - t} secs")
    p.save_index(args.index_path)
    print("saving index to ", args.index_path)